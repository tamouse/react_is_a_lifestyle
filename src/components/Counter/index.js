import React from "react"
import PropTypes from "prop-types"

/*

   Discussion
   ==========

   Counter is a React component class that contains the functionality for
   setting and incrementing a counter, and passing that information onto
   it's children.

   This provides a declartive functionality that can be used in many
   places in our app, while still allowing complete control of display.

   Usage
   -----

       <Counter initialValue={10}>
         {(counter, inc, reset) => {
           return(
             <div>
               <h1>{counter}</h1>
               <button onClick={inc}>add one</button>
               <button onClick={reset}>reset</button>
             </div>
           )
         }
       </Counter>

   Children Function Signature
   ---------------------------

   The function signature for the children render function is:

   function (counterValue, incrementerFunc, resetFunc) {}

   - counterValue [number] current value of counter state
   - incrementFunc [func] function to increment counter, the signuture is:

         function (step) {}

   - resetFunc [func] resets the counter to the initial value property

 */

export default class Counter extends React.Component {
  // ----------------------------------------
  // STATICS
  // ----------------------------------------

  static propTypes = {
    // Allow the caller to set an initial value for the counter
    initialValue: PropTypes.number,

    // Use children as a render prop, taking in the current state of the
    // counter, and the incrementing function.
    children: PropTypes.func.isRequired,

    onChange: PropTypes.func
  }

  static defaultProps = {
    initialValue: 0 /* Start at zero by default */
  }

  // ----------------------------------------
  // INSTANCE VARIABLES
  // ----------------------------------------

  state = {
    _myCounterNotYourCounter: -9999 /* funky long name to indicate that this
                                       state variable is not available to the
                                       outside. */

    /*
       Setting it here to some ridiculous value to show how defaultProps
       works. When the component is first loaded, it will passthis value to
       the children.

       When the lifecycle is complete, the default value has been set in the
       state, and the compoent is rendered again.

       (You don't have to do this in real life)

     */
  }

  // ----------------------------------------
  // INSTANCE METHODS
  // ----------------------------------------

  inc = (step = 1) => {
    // Incrementing function, with a default step of 1
    this.setState(
      state => ({
        _myCounterNotYourCounter: state._myCounterNotYourCounter + step
      }),
      this.afterUpdate
    )
  }

  reset = () => {
    this.setState(
      state => ({
        _myCounterNotYourCounter: this.props.initialValue
      }),
      this.afterUpdate
    )
  }

  afterUpdate = () => {
    // This was the weirdest thing. Apparently, I can't use the name `updater` for a function in a class.
    this.props.onChange &&
      this.props.onChange(this.state._myCounterNotYourCounter)
  }

  // ----------------------------------------
  // LIFECYCLE METHODS
  // ----------------------------------------

  componentDidMount() {
    // Stuff to do when the component is "mounted", i.e., placed into the ShadowDom
    this.setState(state => ({
      _myCounterNotYourCounter: this.props.initialValue
    }))
  }

  render() {
    // Since this is not a display component, there isn't much happening here.
    //
    // What's key, is that we pass the current state and our incrementing
    // function into the children render property function we passed in from
    // above in the component tree.

    const { children } = this.props
    return (
      <React.Fragment>
        {/* a Fragment doesn't add anything to the DOM */}
        {children(this.state._myCounterNotYourCounter, this.inc, this.reset)}
      </React.Fragment>
    )
  }
}
