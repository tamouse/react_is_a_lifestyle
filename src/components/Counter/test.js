import React from "react"
import { shallow } from "enzyme"
import Counter from "./index"

describe("test Counter component", () => {
  let MockChildren, wrapper

  beforeEach(() => {
    MockChildren = jest.fn()
    wrapper = shallow(<Counter children={MockChildren} />)
  })

  it("renders the Counter", () => {
    expect(wrapper).toMatchSnapshot()
  })

  it("calls the children", () => {
    expect(MockChildren).toHaveBeenCalled()
  })

  it("the children are called twice", () => {
    expect(MockChildren.mock.calls.length).toBe(2) // first mount, then lifecycle run
  })

  it("the first time, the counter is set to the ridiculous value from the state definition", () => {
    expect(MockChildren.mock.calls[0][0]).toBe(-9999) // this was the state value for counter
  })

  it("the second time, the counter is set to the default props value", () => {
    expect(MockChildren.mock.calls[1][0]).toBe(0) // after default props applied
  })

  describe("ensure calling the incrementer works", () => {
    let inc

    beforeEach(() => {
      // Let's catch the second prop passed in the call, which should be our incrementor, and run it
      inc = MockChildren.mock.calls[1][1]
    })

    it("Call it without a step", () => {
      inc()
      expect(MockChildren.mock.calls.length).toBe(3) // it should have been called again
      expect(MockChildren.mock.calls[2][0]).toBe(1) // incremented by 1
    })

    it("Call it with a step of 10", () => {
      inc(10)
      expect(MockChildren.mock.calls.length).toBe(3) // it should have been called again
      expect(MockChildren.mock.calls[2][0]).toBe(10) // incremented by 10
    })
  })

  it("ensure reset works", () => {
    const wrapper = shallow(
      <Counter initialValue={7} children={MockChildren} />
    )
    expect(argsFromLastMockCall(MockChildren)[0]).toBe(7)
    const inc = argsFromLastMockCall(MockChildren)[1]
    inc()
    inc()
    inc()
    expect(argsFromLastMockCall(MockChildren)[0]).toBe(10)
    wrapper.instance().reset()
    expect(argsFromLastMockCall(MockChildren)[0]).toBe(7)
  })

  describe("testing Counter onChange ", () => {
    it("calls the onChange callback every time the counter state changes", () => {
      const MockChildren = jest.fn()
      const MockOnChange = jest.fn()
      const wrapper = shallow(
        <Counter children={MockChildren} onChange={MockOnChange} />
      )
      expect(wrapper.instance().props.onChange).toEqual(MockOnChange)
      argsFromLastMockCall(MockChildren)[1]() // inc
      expect(argsFromLastMockCall(MockChildren)[0]).toBe(1)
      expect(MockOnChange).toHaveBeenCalled()
      expect(argsFromLastMockCall(MockOnChange)[0]).toBe(1)
    })
  })
})
