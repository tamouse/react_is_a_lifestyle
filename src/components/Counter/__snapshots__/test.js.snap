// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Counter component renders the Counter 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Counter
    initialValue={0}
>
    [Function]
</Counter>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
